grails-app/jobs/NodeActivatorJob.groovy

```diff
@@ -72,49 +71,58 @@ class NodeActivatorJob {
-            if (log.debugEnabled) {
                 sw?.stop()
                 sw?.start('Finding active node keys')
             }
-            SQLQuery activeNodesQuery = sessionFactory.currentSession.createSQLQuery("""
-                    SELECT n.id FROM process_node n
-                    INNER JOIN basic_process p ON n.process_id = p.id
-                    WHERE p.app_groupid = :appExternalID
-                      AND n.status_id IN ($activatedStatusKey, $runningStatusKey)
-                      AND n.type      IN (:types)
-                    ORDER BY n.started_on ASC""")
-            activeNodesQuery.setParameter('appExternalID', appExternalID)
-            activeNodesQuery.setParameterList('types', [ConstantUtils.NODE_TYPE_ACTIVITY, ConstantUtils.NODE_TYPE_FORK,
-                                                        ConstantUtils.NODE_TYPE_ORJOIN, ConstantUtils.NODE_TYPE_ANDJOIN])
-            List activeNodesKeys = activeNodesQuery.list()
+
+            final Map queryParams = [
+                    appExternalID: appExternalID,
+                    statuses: [
+                            FlowStatus.findByStatusID(NodeStatusEnum.ACTIVATED.value()),
+                            FlowStatus.findByStatusID(NodeStatusEnum.RUNNING.value())
+                    ],
+                    processNodeTypes: [
+                            ConstantUtils.NODE_TYPE_ACTIVITY,
+                            ConstantUtils.NODE_TYPE_FORK,
+                            ConstantUtils.NODE_TYPE_ORJOIN,
+                            ConstantUtils.NODE_TYPE_ANDJOIN
+                    ]
+            ]
+
+            List<Long> activeNodesKeys = ProcessNode.executeQuery("""
+
+                    SELECT pn.id FROM ProcessNode AS pn INNER JOIN pn.process AS process
+                    WHERE process.appGroupID = :appExternalID
+                      AND pn.status IN (:statuses)
+                      AND pn.type IN (:processNodeTypes)
+                    ORDER BY pn.startedOn ASC""",
+
+                    queryParams,
+                    [max: processManagerService.getMaxRunningThreadsQuantity()]
+            )
+
+            Long runningStatusKey = FlowStatus.findByStatusID(NodeStatusEnum.RUNNING.value())?.id
 
             log.info "*** Amount of Nodes to execute ${activeNodesKeys.size()} ***"
 
             if (log.debugEnabled) {
                 sw?.stop()
             }
-            def nodesComparator
+            def nodesComparator = grailsApplication.config.grailsflow.nodeActivator.comparator
 
             if (activeNodesKeys) {
                 if (log.debugEnabled) {
                     sw?.start('Loading ProcessNodes by keys')
                 }
 
-                List activeNodes = []
-                activeNodesKeys.each {
-                    activeNodes << ProcessNode.get(it)
-                }
+                List<ProcessNode> activeNodes = ProcessNode.findAllByIdInList(activeNodesKeys)

```

Comment:
Does it make sense to load IDs first and then load objects? Isn't it better to load objects already in line 92?

Reply:
The previous approach was to get rid of criteria and use plain SQL. But now we have a limit to getting process nodes per one request and we make to try it.